validator/src/bootstrap.rs

```diff
@@ -616,629 +500,297 @@ mod without_incremental_snapshots {
+                if rpc_peers.len() > 1 { "s" } else { "" },
+                rpc_peers,
+            );
+
+            return Some(GetRpcNodeResult {
+                rpc_contact_info: final_rpc_contact_info,
+                snapshot_hash: Some(final_snapshot_hash),
+            });
         }
     }
+}
 
-    fn get_known_snapshot_hashes(
-        cluster_info: &ClusterInfo,
-        known_validators: &Option<HashSet<Pubkey>>,
-    ) -> Option<HashSet<(Slot, Hash)>> {
-        if let Some(known_validators) = known_validators {
-            let mut known_snapshot_hashes = HashSet::new();
-            for known_validator in known_validators {
-                cluster_info.get_snapshot_hash_for_node(known_validator, |snapshot_hashes| {
-                    for snapshot_hash in snapshot_hashes {
-                        known_snapshot_hashes.insert(*snapshot_hash);
-                    }
-                });
+/// Get the Slot and Hash of the local snapshot with the highest slot.  Can be either a full
+/// snapshot or an incremental snapshot.
+fn get_highest_local_snapshot_hash(
+    full_snapshot_archives_dir: impl AsRef<Path>,
+    incremental_snapshot_archives_dir: impl AsRef<Path>,
+    incremental_snapshot_fetch: bool,
+) -> Option<(Slot, Hash)> {
+    snapshot_utils::get_highest_full_snapshot_archive_info(full_snapshot_archives_dir).and_then(
+        |full_snapshot_info| {
+            if incremental_snapshot_fetch {
+                snapshot_utils::get_highest_incremental_snapshot_archive_info(
+                    incremental_snapshot_archives_dir,
+                    full_snapshot_info.slot(),
+                )
+                .map(|incremental_snapshot_info| {
+                    (
+                        incremental_snapshot_info.slot(),
+                        *incremental_snapshot_info.hash(),
+                    )
+                })
+            } else {
+                None
             }
-            Some(known_snapshot_hashes)
-        } else {
-            None
-        }
-    }
+            .or_else(|| Some((full_snapshot_info.slot(), *full_snapshot_info.hash())))
+        },
+    )
+}
 
-    /// Get the Slot and Hash of the local snapshot with the highest slot.
-    fn get_highest_local_snapshot_hash(
-        full_snapshot_archives_dir: impl AsRef<Path>,
-    ) -> Option<(Slot, Hash)> {
-        snapshot_utils::get_highest_full_snapshot_archive_info(full_snapshot_archives_dir)
-            .map(|full_snapshot_info| (full_snapshot_info.slot(), *full_snapshot_info.hash()))
+/// Get peer snapshot hashes
+///
+/// The result is a vector of peers with snapshot hashes that:
+/// 1. match a snapshot hash from the known validators
+/// 2. have the highest incremental snapshot slot
+/// 3. have the highest full snapshot slot of (2)
+fn get_peer_snapshot_hashes(
+    cluster_info: &ClusterInfo,
+    validator_config: &ValidatorConfig,
+    rpc_peers: &[ContactInfo],
+    incremental_snapshot_fetch: bool,
```

Comment:
Minor/Preference: Same comment here w.r.t. passing in `bootstrap_config` instead of the `incremental_snapshot_fetch` flag. (And above `rpc_peers`/below `validator_config` for consistency.)